### DISCUSSION TITLE
DID 2612, 2613 —> Fan-ODU

### DISCUSSION THREAD

**Discussion by bu-na** [open3e/open3e#142] [Q&A] [answered]
Hier ein Vorschlag zur Benennung von 2612 und 2613 zur Diskussion:

$ open3e  -r 0x68c.2612,0x68c.2613 -v
0x68c 2612 PrimarySourceCommonSettingsHeating "00e8038a02c800"
0x68c 2613 PrimarySourceCommonSettingsCooling "00e8038a02c800"
closing 0x680 - bye!
closing 0x68c - bye!

2612 : O3EComplexType(7, "PrimarySourceCommonSettingsHeating", [O3EByteVal(1, "Mode"), O3EInt16(2, "MaxFanSpeed"), O3EInt16(2, "DefaultFanSpeed"), O3EInt16(2, "MinFanSpeed")]),# ODU-FanSpeed --> Mode: Variable/Fixed, Unit: %
2613 : O3EComplexType(7, "PrimarySourceCommonSettingsCooling", [O3EByteVal(1, "Mode"), O3EInt16(2, "MaxFanSpeed"), O3EInt16(2, "DefaultFanSpeed"), O3EInt16(2, "MinFanSpeed")]),# ODU-FanSpeed --> Mode: Variable/Fixed, Unit: %

Gruß Bu-Na

**Comment by Philip-Wiege** [comment]
Hi!

Wo hast du denn die Infos für die Dekodierung her? Hast du das selber rausgeforscht oder eine andere Quelle?

**Reply by bu-na** [reply]
siehe VS-Community Waermepumpe-Hybridsysteme/Vitocal-250-A-Energiebilanz

**Comment by Philip-Wiege** [comment, marked as answer]
Ich hab den Änderungsvorschlag mal kurz lokal in mein open3e eingebaut. Bei mir kommen 57% max Drehzahl und 10% min Drehzahl raus. Das deckt sich mit meinen Aufzeichnungen aus HomeAssistant. Da werden diese Grenzen tatsächlich eingehalten. Die Drehzahlen würde ich bei mir wirklich gerne hochschrauben zur COP Optimierung. Wenn die WP mal aufdreht hört man bei uns eigentlich immer nur den Verdichter summen. Den Lüfter habe ich noch nie wahrgenommen.

@MyHomeMyData @bu-na Ich würde mal versuchen, meinen ersten Pull-Request zu machen :) Oder willst du @bu-na das machen?

**Reply by bu-na** [reply]
@Philip-Wiege mach mal bitte -:)…

**Reply by Philip-Wiege** [reply]
@MyHomeMyData @philippoo66 Kann es sein, dass ich keine Schreibrechte auf den Develop Branch habe? Ich muss doch erst dort meinen lokalen Commit hin "pushen" oder? Der Pull Request käme dann erst, wenn das Ganze vom Develop in den Master soll oder?

**Reply by MyHomeMyData** [reply]
@Philip-Wiege Stimmt, Du hattest keine Schreibrechte. Habe ich eben geändert. Bitte nochmal probieren.

**Comment by Philip-Wiege** [comment]
Hi @bu-na, könntest du mir bei der Parametrierung meiner Lüfterdrehzahlen helfen?

Wenn ich 2612 raw auslese bekomme ich:

`raspiopen3e@IT-Open-3E:~/open3e $ python3 Open3Eclient.py -c can0 -r 0x68c.2612 -raw

0x68c 013a023a026400`

Das entspricht 01, 57, 57, 10 wenn ich little endian decodiere mit uint16 faktor 10.
Jetzt versuche ich 01, 60, 57, 10 zurückzuschreiben. ich habe nur das 3A02 durch 5802 ersetzt für die Max-Speed:

`raspiopen3e@IT-Open-3E:~/open3e $ python3 Open3Eclient.py -c can0 -tx 0x68C -raw -w 2612=0158023A026400

write raw: 1676.2612 = 0158023A026400
Traceback (most recent call last):
  File "/home/raspiopen3e/open3e/Open3Eclient.py", line 380, in <module>
    succ,code = dicEcus[ecu].writeByDid(didkey, didVal, raw=True)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/raspiopen3e/open3e/Open3Eclass.py", line 147, in writeByDid
    response = self.uds_client.write_data_by_identifier(did, val)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/raspiopen3e/.local/lib/python3.11/site-packages/udsoncan/client.py", line 167, in decorated
    return func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/raspiopen3e/.local/lib/python3.11/site-packages/udsoncan/client.py", line 521, in write_data_by_identifier
    response = self.send_request(req)
               ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/raspiopen3e/.local/lib/python3.11/site-packages/udsoncan/client.py", line 2255, in send_request
    raise NegativeResponseException(response)
udsoncan.exceptions.NegativeResponseException: WriteDataByIdentifier service execution returned a negative response ConditionsNotCorrect (0x22)`

Musstest du für das Schreiben noch Vorbedingungen erfüllen oder ist der Datenpunkt in der VCMU (0x68C) ein geschützter Datenpunkt?


@MyHomeMyData Weißt du, ob man mit Service 77 auf dem internen Bus auch die VCMU erreichen kann?

**Reply by bu-na** [reply]
Habe mal probiert, den Parameter zu schreiben, gleicher Effekt. Scheint am externen CAN-Bus auf ReadOnly zu stehen. 
Im Unterschied zu dir steht der Mode bei meinem System auf "0"
(0x68c 2612 PrimarySourceCommonSettingsHeating "00e8038a02c800"). Möglicherweise spielt das auch irgendwo eine Rolle.
Der nächste Versuch wäre dann F77 am internen Bus zu testen.

**Comment by Philip-Wiege** [comment]
Ich habe gerade noch mal versucht, die Lüfterdrehzahl mit dem Service 77 auf dem internen Bus zu schreiben. Leider gabs einen Fehler. Ich habe den 149er Branch verwendet.

```
(.venv) raspiopen3e@IT-Open-3E:~/open3e_pip_149 $ open3e -c can0 -w 0x68C.2612="013a023a026400" -v -raw -f77
Configuration file devices.json not found, continuing with generic DID list, ECU 0x680
hallo! 149
write raw: 1676.2612 = 013a023a026400
Traceback (most recent call last):
  File "/home/raspiopen3e/open3e_pip_149/.venv/bin/open3e", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/raspiopen3e/open3e_pip_149/.venv/lib/python3.11/site-packages/open3e/Open3Eclient.py", line 475, in main
    ecu77 = open3e.Open3Eclass.O3Eclass(ecutx=dicEcus[ecu].tx+2, doip=args.doip, can=args.can, dev=args.dev)
                                              ~~~~~~~^^^^^
KeyError: 1676
```

@MyHomeMyData @philippoo66  Irgendeine Ahnung, was da passiert? Kann es sein, dass sich da meine ECU Adressierung beißt?

**Reply by philippoo66** [reply]
oh, da hat er die ECU als DID interpretiert glaub ich. hast du den Output zufälling noch mal als screenshot?

edit: hab's in deinem Beitrag geändert, sry. Bei sowas machen sich 3 ` besser als einer

**Reply by philippoo66** [reply]
so sieht es schon anders aus. irgendwas hat mit dem ensure_ecu() nicht hingehauen. aber ich muss es trotzdem selber mal laufen lassen glaubich...

es sieht so aus, als würde das ensure_ecu() bei explizitem 77 garnicht aufgerufen

**Reply by MyHomeMyData** [reply]
@philippoo66 Der wesentliche Unterschied bei Verwendung von -f77 liegt ja in der Erzeugung eines neuen ECUClients, also in der Zeile
```
ecu77 = open3e.Open3Eclass.O3Eclass(ecutx=dicEcus[ecu].tx+2, doip=args.doip, can=args.can, dev=args.dev)
```
Ich blicke da zu wenig durch, aber vielleicht ist die Generierung der ecutx-Adresse das Problem.

**Reply by philippoo66** [reply]
@Philip-Wiege  ok, ich hab's 'dazwischen gekloppt' ohne zu testen (muss erstmal Kaffee trinken). Wenn du schneller bist kannst du es jetzt noch mal versuchen

**Reply by philippoo66** [reply]
@MyHomeMyData  die 77er ECU machst du ja selber. aber er greift dabei ja auf die ECU mit der eigentlichen txID in dem dictionary zu. ohne ensure_ecu gibt es da aber ohne devices.json erstmal nur die 0x680. Also habe ich wie bei allen anderen Aufrufen das ensure_ecu einfach eine (2) Zeilen davor eingefügt. So müsste es gehen so sauber wie wir programmiert haben :-)

**Reply by Philip-Wiege** [reply]
Kann es sein, dass es noch mehr Stellen im Code gibt, wo das ensure ECU rein muss? SID 77 wird ja an mehr Stellen aufgerufen und mein Fehler tritt in Zeile 475 auf. Du hattest an einer anderen Stelle hinzugefügt. Wenn du das einbaust, kann ich gleich testen.

**Reply by philippoo66** [reply]
man hääte es glaubich auch anders machen können per 
`ecu77 = open3e.Open3Eclass.O3Eclass(ecutx=ecu+2, doip=args.doip, can=args.can, dev=args.dev)`
aber das mit dem ensure kommt überall, also ist es mit 'geradliniger'

**Reply by philippoo66** [reply]
@Philip-Wiege  oh du hast recht. jetzt hab ich es da auch noch gemacht. fehlt es sonst noch wo? ich schau auch noch mal...

**Reply by MyHomeMyData** [reply]
Das muss dann im MQTT-Zweig, also ab Zeile 243 auch rein, oder?

**Reply by philippoo66** [reply]
da steht es ja schon vor der for schleife (Z 242)

**Reply by philippoo66** [reply]
ich hab noch ein drittes f77 gefunden wo es nicht bei war. ich hoffe jetzt hab ich alle

**Reply by philippoo66** [reply]
@MyHomeMyData  in Z.243ff hast du es ja sogar per `ecutx=addr+2` gemacht. da wäre es wahrscheinlich garnicht nötig gewesen. Aber k.A. ob irgendwo (z.B. bei der Anzeige/Ausgabe) noch auf das dicEcus zugegriffen wird...

**Reply by MyHomeMyData** [reply]
@philippoo66 Pass das gerne so an, wie Du es machen würdest.

**Reply by philippoo66** [reply]
ok, aber da muss ich jetzt erstmal Kaffee trinken. in den meisten Fällen ist ja die ECU gleich der Adresse, nur wenn sie in der devices.json 'sprechend' (um)benannt wurde, dann braucht es dieses dict meine ich. Und ob das mit "0x6yz" immer hinhaut ohne get_int()?...

**Reply by philippoo66** [reply]
Es wird im gesamten Code einfach davon ausgegangen, dass eine benutzte ECU auch in dem dictEcus existiert (abgesehen von den nur temporär existierenden 77er), und das soll auch so bleiben, dann muss man sich da keine Gedanken machen.

**Reply by philippoo66** [reply]
ok, ich habe es jetzt alles in `ecutx=ecu+2` geändert. steht überall `ecu = lsteds[0][0]  # int` drüber, sollte als gefahrlos sein. dann ist es gleich wie bei `ecutx=addr+2`. ist gradliniger und jetzt auch durchgängig. das ensure_ecu() hab ich trotzdem stehen lassen aus o.a. Gründen.

**Reply by Philip-Wiege** [reply]
Also es scheint jetzt zu laufen :) Der Request wird glaube ich rausgeschickt. Leider kommt aber ein Timeout :( Irgendwie muss Viessmann ja bei @bu-na auch die Drehzahlen bedatet haben....

`(.venv) raspiopen3e@IT-Open-3E:~/open3e_pip_149 $ open3e -c can0 -w 0x68C.2612="013a023a026400" -v -raw -f77
Configuration file devices.json not found, continuing with generic DID list, ECU 0x680
hallo! 149
write raw: 1676.2612 = 013a023a026400
Using writeDataByIdentifier service 77. Verify the result!
<Request: [WriteDataByIdentifier77] - 15 data bytes at 0x7fa473acd0>
return: TimeoutException, code: Did not receive response in time. Global request timeout time has expired (timeout=3.000 sec)
closing 0x680 - bye!
^[[Aclosing 0x68c - bye!`

**Reply by Philip-Wiege** [reply]
@philippoo66 Dein zusätzliches print kommt glaube ich nicht durch. In der Konsole unten steht, dass 0x68C geschlossen wird. Kann es sein, dass die ECU Verbindung 0x68E (0x68C+2) garnicht verwendet wird? Läuft der Service 77 auf der normalen ECU Adresse einfach ins Leere?

Grüße
Philip

```
(.venv) raspiopen3e@IT-Open-3E:~/open3e_pip_149 $ open3e -c can0 -w 0x68C.2612="013a023a026400" -v -raw -f77
Configuration file devices.json not found, continuing with generic DID list, ECU 0x680
hallo! 149
write raw f77: 1676.2612 = 013a023a026400
Using writeDataByIdentifier service 77. Verify the result!
<Request: [WriteDataByIdentifier77] - 15 data bytes at 0x7faba6add0>
return: TimeoutException, code: Did not receive response in time. Global request timeout time has expired (timeout=3.000 sec)
closing 0x680 - bye!
closing 0x68c - bye!
```

**Reply by philippoo66** [reply]
Moin Philip!
Das 'generalisierte' print steht in der 4. Zeile deiner Ausgabe wie es sein soll.
Die ecu77 Instanz wird nur temporär für den ein-mal Zugriff geschaffen und wird direkt danach wieder geschlossen, zugegeben ohne die 'bye!' Meldung, die ja ganz unten im main erfolgt. Sie wird aber definitiv benutzt, wie die Ausgaben Zeilen 5 und 6 'dokumentieren'.
Meist du es wäre sinnvoll, die 'bye!' Message in die class zu verlegen? (bzw. aus dem cliennt heraus auszugeben - die class soll eigentlich kein print machen)

ps. wenn du bei dem hier-rein-Kopieren der Ausgaben einen 3-fachen 'Rückwärts-Akzent' ( ` ) benutzt um den paste einzufassen, brauch ich nicht immer deine Nachrichten editieren, um das in übersichtliche Form zu bringen ;-) 

![grafik](https://github.com/user-attachments/assets/78dbb4c3-fd2a-49a8-869d-42bfc01e5fbd)

**Reply by philippoo66** [reply]
pps. das print gibt immer die 'originale' ECU-tx aus. Wie du aber selber in deinem candump gesehen hast, wird der Service 77 Request unter 68e rausgeschickt. Eventuell macht es Sinn, das print mit der +2 COB auszugeben?!? Das könnte aber auch wieder Fragen beim 'Normalbenutzer' aufwerfen, dem das +2 garnicht gewahr ist
